kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/Metadata.kt

```diff
@@ -0,0 +1,350 @@
+
+  override fun visitAbbreviatedType(flags: Flags): KmTypeVisitor? {
+    if (!useTypeAlias) {
+      return null
+    }
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      typeAliasType = it
+    }
+  }
+
+  override fun visitArgument(flags: Flags, variance: KmVariance): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      argumentList.add(
+          when (variance) {
+            KmVariance.IN -> WildcardTypeName.supertypeOf(it)
+            KmVariance.OUT -> {
+              if (it == ANY) {
+                // This becomes a *, which we actually don't want here.
+                // List<Any> works with List<*>, but List<*> doesn't work with List<Any>
+                it
+              } else {
+                WildcardTypeName.subtypeOf(it)
+              }
+            }
+            KmVariance.INVARIANT -> it
+          }
+      )
+    }
+  }
+
+  override fun visitClass(name: kotlinx.metadata.ClassName) {
+    className = name
+  }
+
+  override fun visitFlexibleTypeUpperBound(flags: Flags,
+      typeFlexibilityId: String?): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      flexibleTypeUpperBound = WildcardTypeName.subtypeOf(it)
+    }
+  }
+
+  override fun visitOuterType(flags: Flags): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      outerType = WildcardTypeName.supertypeOf(it)
+    }
+  }
+
+  override fun visitStarProjection() {
+    argumentList.add(WildcardTypeName.STAR)
+  }
+
+  override fun visitTypeAlias(name: kotlinx.metadata.ClassName) {
+    if (!useTypeAlias) {
+      return
+    }
+    typeAliasName = name
+  }
+
+  override fun visitTypeParameter(id: Int) {
+    typeParameter = getTypeParameter?.invoke(id) ?: throw IllegalStateException(
+        "Visiting TypeParameter when there are no type parameters!")
+  }
+
+  override fun visitEnd() {
+    var finalType = flexibleTypeUpperBound ?: outerType ?: typeParameter
+    if (finalType == null) {
+      if (useTypeAlias) {
+        finalType = typeAliasType ?: typeAliasName?.let { ClassName.bestGuess(it.replace("/", ".")) }
+      }
+      if (finalType == null) {
+        finalType = className?.let { ClassName.bestGuess(it.replace("/", ".")) }
+            ?: throw IllegalStateException("No valid typename found!")
+        if (argumentList.isNotEmpty()) {
+          finalType = (finalType as ClassName).parameterizedBy(*argumentList.toTypedArray())
+        }
+      }
+    }
+
+    finalType = finalType.asNullableIf(nullable)
+    receiver(finalType)
+  }
+}
+
+internal fun KotlinClassMetadata.Class.readClassData(): ClassData {
+  @Suppress("RedundantExplicitType")
+  var classFlags: Flags = 0
+  lateinit var className: String
+  var constructorData: ConstructorData? = null
+  var companionObjectName: String? = null
+  val typeParameters = LinkedHashMap<Int, TypeVariableName>()
```

Comment:
mapOf ?

Reply:
it technically uses `LinkedHashMap` under the hood, but that's an implementation detail and we care about maintaining order for idempotence. This just makes that explicit. Open to suggestions though if you had something more idiomatic in mind!